### ISSUE TITLE
Can Laya be used as a local decision model for fast-jev-compaction-style tool-call compaction?

### ISSUE THREAD

**Issue by vanlin-ali** [NandhaKishorM/laya#44]
# Can Laya be used as a local decision model for fast-jev-compaction-style tool-call compaction?

## Context

I'm evaluating [Laya](https://github.com/NandhaKishorM/laya) as a local decision model for a tool-call compaction workflow similar to [fast-jev-compaction](https://github.com/tamaratran/fast-jev-compaction).

The target is not generic text classification. The concrete decision problem is:

1. **Should this tool call be kept?**
2. **Should this tool result be kept verbatim?**

The resulting action is:

- keep call + keep result
- keep call + truncate/drop result
- drop call + drop result

This is important because `fast-jev-compaction` uses Jev `noul` questions to make these decisions and applies a threshold to the returned scores.

## Reference workflow

The `fast-jev-compaction` implementation works roughly as follows:

1. Pair each `tool_use` with its corresponding `tool_result`.
2. Keep pinned messages and the most recent messages.
3. Build the state presented to the decision model.
4. Ask decision questions for each non-pinned tool call:
   - whether the call should remain
   - whether the result should remain verbatim
5. Apply `keepThreshold`:
   - `keepResult >= threshold` → keep call + result
   - otherwise, if `keepCall >= threshold` → keep call and truncate the result
   - otherwise → remove call + result
6. Rebuild the conversation without rewriting retained content.

Therefore, the useful property of a replacement model is not simply that its scores are numerically calibrated. The scores need to have useful **discrimination / ranking** for the actual keep/drop decisions.

## Experiment

I evaluated Laya on four real tool-call examples from the compaction workflow.

| Case | Tool | Jev score | Jev decision | Calibrated Laya score | Our decision |
|---|---|---:|---|---:|---|
| t1 | Read | 0.18 | `drop_call` | 0.687 | keep |
| t2 | Bash | 0.64 | `drop_result` | 0.647 | keep |
| t3 | Grep | 0.22 | `drop_call` | 0.602 | keep |
| t4 | Edit | 0.71 | `keep` | 0.537 | keep |

Calibration itself appears to work numerically:

- Before calibration, Laya scores were approximately `0.86–0.99`.
- After calibration, they were compressed to approximately `0.54–0.69`.

However, the actual decisions did not improve.

### Decision agreement

Using the final Jev decision as the target:

- Agreement: `1 / 4 = 25%`
- This is the same as an always-`keep` baseline on this sample.

More importantly, the ranking appears to be reversed.

The calibrated Laya scores are:

```text
t1 Read   = 0.687
t2 Bash   = 0.647
t3 Grep   = 0.602
t4 Edit   = 0.537

**Comment by ThinkyMiner** [comment]
been looking at this, i think a few things in the setup are off before we can say laya can't do it.

the calibration part first. temperature scaling just divides the logits by one scalar, so it's monotonic. it can squeeze 0.86-0.99 down into 0.54-0.69 but it can't swap two items around. so the ranking staying the same after calibration is expected, nothing went wrong there. what compaction actually needs is discrimination (higher score = more keepable) and calibration doesn't touch that at all.

second, which checkpoint did you load? the 0.766 in the readme is laya-typed-decisions. the base ones score 0.362 and 0.342 on that same benchmark, which is under the 0.461 majority-class baseline. readme says it straight out, all the capability comes from fine-tuning. so if you used convaiinnovations/laya you basically measured a chance-level model on 4 examples and got a reversed order, which is about what you'd expect. try subfolder="typed-decisions", it has agent_trace_observability at 0.730 which is the closest thing to compaction in its training data.

third thing, and this one isn't documented anywhere so easy to miss: noul takes a criteria dict. if you leave it out, render_options falls back to the literal strings "false: no, the statement does not hold" / "true: yes, the statement holds", and those are what actually go into the sequence behind the mask tokens. so the model is comparing your tool result against two placeholders.

and honestly the bigger issue might be the state. keepability isn't a property of the call on its own. a Read is droppable because the file got edited later, a Grep is droppable because you already used the answer. if you're only passing the call + result then that info isn't in the input and no model is getting it right. something like:

state = {
    "tool": "Read",
    "args": {"file_path": "src/auth.py"},
    "result_preview": result[:1200],   # ~768 tokens of state budget at max_len 1024
    "turns_since": 14,
    "superseded_by_later_call": True,
    "current_goal": "fix the failing test in tests/test_router.py",
}

also 4 examples measured as agreement can't tell you whether there's no signal or just a bad threshold, both look the same. auc against the jev action would separate those, and it doesn't depend on where keepThreshold sits.

if it does turn out there's no signal, the kaggle notebook is basically ready to go. point MODEL_ID at the typed-decisions subfolder and swap the dataset in cell 6 for your own logged decisions, using jev's score as a soft target {"true": jev_score, "false": 1 - jev_score}. the rlcd reward eats full distributions so soft targets are better than hardening them to 0/1. rest of it runs as is, 4-5 hours on the free 2xT4.

i'd frame the whole thing as distillation rather than replacement tbh. you're not trying to beat jev, just match its ranking well enough that the same threshold makes the same calls, at 33ms and no per-token cost. seems doable for a 421M encoder but it needs the logged data, it's not going to work zero-shot and the readme is upfront about that.
